src/plugins/shared-variables/slate/variables-plugin.tsx

```diff
@@ -3,75 +3,141 @@ import classNames from "classnames/dedupe";
+      if (!sharedModelManager || !sharedModelManager.isReady) {
+        // So we need to keep waiting until the sharedModelManager is ready
+        if (DEBUG_SHARED_MODELS) {
+          console.log("shared model manager isn't available");
+        }
+        return;
+      }
+
+      if (this.sharedModel) {
+        // We already have a shared model so we don't need to do anything
+        return;
+      }
+
+      // We don't have a shared model, see if the there a SharedVariables model
+      const containerSharedModel = sharedModelManager.findFirstSharedModelByType(SharedVariables);
+      if (!containerSharedModel) {
+        if (DEBUG_SHARED_MODELS) {
+          console.log("no shared variables model in the document");
+        }
+
+        // If we want to automatically create a shared model this would be the
+        // place to put it.
+        return;
+      }
+
+      // We found a SharedVariables model, and we don't have one on the textContent yet
+      // So add it
+      sharedModelManager.addTileSharedModel(this.textContent, containerSharedModel);
 
-export const insertTextVariable = (variable: VariableType, editor?: Editor) => {
-  if (!editor) {
-    console.warn("inserting variable but there is no editor");
-    return;
+      // We could dispose the autorun here, but we aren't. There is a chance
+      // that the shared model will be removed from the document and a new one
+      // added. This is not currently supported, but it seems pretty harmless to
+      // leave the autorun in place just in case we support this later.
+    });
+  }
+
+  onInitEditor(editor: CustomEditor) {
+    return withVariables(editor, this.textContent);
+  }
+
+  dispose() {
+    this.disposeSharedModelManagerAutorun?.();
   }
-  const reference = variable.id;
-  const varElt: VariableElement = { type: kVariableFormat, reference, children: [{text: "" }]};
-  Transforms.insertNodes(editor, varElt);
-};
 
-export const findSelectedVariable = (selectedElements: any, variables: VariableType[]) => {
-  let selected = undefined;
-    // FIXME: The editor.selectedElements claims that it returns a BaseElement[] but it really returns a NodeEntry
-    // which is a list of pairs. [Node, Path]
-    // https://docs.slatejs.org/api/nodes/node-entry
-    // There's some weirdness below to work around that, but
-    // we should either update the return type our slate lib or just return the BaseElement list.
-  selectedElements?.forEach((selectedItem: any) => {
-    const baseElement = (selectedItem as any)[0];
-    if (isVariableElement(baseElement)) {
-      const {reference} = baseElement;
-      selected = variables.find(v => v.id === reference);
+  get chipVariables() {
+    const {editor} = this.textContent;
+    const variableIds: string[] = [];
+    if (editor) {
+      for (const [node] of Editor.nodes(editor, {at: [], mode: 'all'})) {
+        if (Editor.isInline(editor, node) && isVariableElement(node)) {
```

Comment:
```suggestion
        if (isVariableElement(node)) {
```
Seems like this should be sufficient since all variables nodes are inline elements.